kit/scrolls/src/Lists/UICollapsableSectionList/UICollapsableSectionList.tsx

```diff
@@ -0,0 +1,323 @@
+/* eslint-disable no-param-reassign */
+/* eslint-disable no-underscore-dangle */
+import * as React from 'react';
+import {
+    DefaultSectionT,
+    SectionListData,
+    SectionListProps,
+    TouchableOpacity,
+    StyleSheet,
+} from 'react-native';
+// TODO: it won't work on web. Consider to do sth with it, when do implementation for web.
+// @ts-ignore
+import VirtualizedSectionList from 'react-native/Libraries/Lists/VirtualizedSectionList';
+
+import { AccordionOverlayView, AccordionOverlayViewRef } from './AccordionOverlayView';
+import {
+    useVirtualizedListFramesListener,
+    VirtualizedListFrame,
+    VirtualizedListScrollMetrics,
+} from './useVirtualizedListFramesListener';
+
+let now: number;
+
+const emptyArray: any = [];
+const LAST_SECTION_TAG = 'LAST_SECTION_TAG_DO_NOT_USE_THIS_EXTERNALLY';
+const duration = 1000;
+
+/**
+ * TODO: known problems.
+ * This is was caught on Android:
+ * - (minor and I don't want to address it right now)
+ *   If one tries to expand a section that is the last one
+ *   it won't change the position of the section,
+ *   therefore there wouldn't be any visual feedback
+ *   that it's expanded and one have to scroll manually
+ *   to see some items in the expanded section
+ */
+
+async function prepareAnimation<ItemT, SectionT = DefaultSectionT>(
+    sectionKey: string,
+    foldedSections: Record<string, boolean>,
+    screenshotRef: { current: AccordionOverlayViewRef | null },
+    listRef: { current: VirtualizedSectionList<ItemT, SectionT> },
+    sectionsMapping: { current: Record<string, string> },
+    sectionToAnimateKey: { current: string | undefined },
+) {
+    const list = listRef.current.getListRef();
+
+    sectionToAnimateKey.current = sectionsMapping.current[sectionKey];
+    const currentSectionFrame: VirtualizedListFrame = list._frames[sectionKey];
+    const { visibleLength, offset, contentLength } =
+        list._scrollMetrics as VirtualizedListScrollMetrics;
+    const realBottomOffset = Math.min(offset + visibleLength, contentLength);
+    const visibleBottomOffset = offset + visibleLength;
+    const sectionEndY = currentSectionFrame.offset + currentSectionFrame.length;
+    const isFolded = foldedSections[sectionKey];
+
+    if (sectionToAnimateKey.current !== LAST_SECTION_TAG) {
+        /**
+         * Just show a screenshot above
+         * Animation is handled later in frame change listener
+         */
+        screenshotRef.current?.show(sectionEndY, sectionEndY + visibleLength);
```

Comment:
But what if `hide` won't happen in the frame listener (you've checked only three cases, but what to do for the rest? aren't they possible for 100%?)

Reply:
I hope that I handle all of them :)